adapter/redisQuota/redisQuota.go

```diff
@@ -0,0 +1,326 @@
+func newAspectWithDedup(env adapter.Env, ticker *time.Ticker, c *config.Params, definitions map[string]*adapter.QuotaDefinition) (adapter.QuotasAspect, error) {
+	connPool, _ := NewconnPoolImpl(c.RedisSocketType, c.RedisUrl, c.RedisPoolSize)
+	rq := &redisQuota{
+		definitions: definitions,
+		cells:       make(map[string]int64),
+		redisPool:   connPool,
+		recentDedup: make(map[string]dedupState),
+		oldDedup:    make(map[string]dedupState),
+		ticker:      ticker,
+		logger:      env.Logger(),
+	}
+	return rq, nil
+}
+
+func (rq *redisQuota) Alloc(args adapter.QuotaArgs) (adapter.QuotaResult, error) {
+	return rq.alloc(args, false)
+}
+
+func (rq *redisQuota) AllocBestEffort(args adapter.QuotaArgs) (adapter.QuotaResult, error) {
+	return rq.alloc(args, true)
+}
+
+func (rq *redisQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (adapter.QuotaResult, error) {
+	amount, exp, err := rq.commonWrapper(args, func(d *adapter.QuotaDefinition, key string, currentTime time.Time, currentTick int64) (int64, time.Time,
+		time.Duration) {
+		result := args.QuotaAmount
+		conn, err := rq.redisPool.Get()
+		if err != nil {
+			return 0, time.Time{}, 0
+		}
+		defer rq.redisPool.Put(conn)
+
+		// increase the value of this key by the amount of result
+		conn.PipeAppend("INCRBY", key, result)
+		resp, err := conn.PipeResponse()
+
+		if err != nil {
+			return 0, time.Time{}, 0
+		}
+		ret := int64(resp.Int())
+
+		if ret > d.MaxAmount {
+			if !bestEffort {
+				return 0, time.Time{}, 0
+			}
+			// grab as much as we can
+			result = d.MaxAmount - (ret - result)
+		}
+
+		return result, currentTime.Add(d.Expiration), d.Expiration
+	})
+
+	return adapter.QuotaResult{
+		Amount:     amount,
+		Expiration: exp,
+	}, err
+
+}
+
+func (rq *redisQuota) ReleaseBestEffort(args adapter.QuotaArgs) (int64, error) {
+	amount, _, err := rq.commonWrapper(args,
+		func(d *adapter.QuotaDefinition, key string, currentTime time.Time, currentTick int64) (int64, time.Time, time.Duration) {
+			result := args.QuotaAmount
+			conn, err := rq.redisPool.Get()
+			if err != nil {
+				return 0, time.Time{}, 0
+			}
+			defer rq.redisPool.Put(conn)
+
+			// decrease the value of this key by the amount of result
+			conn.PipeAppend("DECRBY", key, result)
+			resp, err := conn.PipeResponse()
+			if err != nil {
+				return 0, time.Time{}, 0
+			}
+			ret := int64(resp.Int())
+
+			if ret <= 0 {
+				// delete the key since it contains no useful state
+				conn.PipeAppend("DEL", key)
+				// consume the output of previous command
+				conn.PipeResponse()
+			}
+
+			return result, time.Time{}, 0
+		})
+
+	return amount, err
+}
+
+func (rq *redisQuota) Close() error {
+	return nil
+}
+
+type quotaFunc func(d *adapter.QuotaDefinition, key string, currentTime time.Time, currentTick int64) (int64, time.Time, time.Duration)
+
+// TODO: extract the common code below between memQuota and redisQuota into a util-like package, need to figure out where to put the package?
```

Comment:
Note that we have the same sharing issue around tests. Many of the tests for memQuota would apply equally well to redisQuota. We just need to refactor things to enable this.

In fact, I opened issue #436 this morning to track the creation of generic aspect tests for all the different aspect types.